### PR TITLE
Pull #12316: add release number validation

### PR DESCRIPTION
**Pull request by stoyanK7**
As requested in https://github.com/checkstyle/contribution/pull/670#issuecomment-1285754978

### CODE REVIEW

File: .ci/releasenotes-gen.sh
Comment (rnveach) [inline .ci/releasenotes-gen.sh:20]: Assume this is just for testing.

Comment (stoyanK7) [reply .ci/releasenotes-gen.sh:20]: Correct, it is.

**Comment by romani** [comment]
Execution was skipped
https://cloud.drone.io/checkstyle/checkstyle/13029

For testing purposes , please remove skip for PRs, to see execution

**Comment by romani** [comment]
CI did make execution.

>fatal: Could not read from remote repository.

But ok, functional PR is merged.
Please use main repo but please keep removal of execution skip on PR.
Make it run.
Wait for drone to execute. As you get results.
Share link here as comment.
Update commit to be ready for merge.

**Comment by romani** [comment]
@stoyanK7 , it is not failing https://cloud.drone.io/checkstyle/checkstyle/13052/4/2
please investigate.
please create your personal Github token and run releasenotes from your local to investgate and debug a problem.

**Comment by romani** [comment]
GitHub, rebase

**Comment by rnveach** [comment]
@romani We should consider the rebase action broken.
https://github.com/checkstyle/checkstyle/actions/runs/3325206341/jobs/5497637360
> fatal: detected dubious ownership in repository at '/github/workspace'
To add an exception for this directory, call:
	git config --global --add safe.directory /github/workspace

**Comment by romani** [comment]
@stoyanK7 , please add pull number to commit to keep reference to all reasons and discussions.

**Comment by stoyanK7** [comment]
@romani Done. Though there is still a small problem. The error message is not printed: https://cloud.drone.io/checkstyle/checkstyle/13134/4/2. Shall I open another Pull request in `contribution` for that?

**Comment by stoyanK7** [comment]
This one is good to go now I think.

**Comment by rnveach** [comment]
@stoyanK7 Please rebase this on the latest master.

**Comment by stoyanK7** [comment]
[Drone](https://cloud.drone.io/checkstyle/checkstyle/13233/4/2)
```
Generation ends with 1 errors.


[ERROR] Validation of release number failed. Release number is minor(10.5), but release notes do not contain 'new' or 'breaking compatability' labels. Please correct release number by running https://github.com/checkstyle/checkstyle/actions/workflows/bump-version-and-update-milestone.yml
```

**Comment by romani** [comment]
Heh, this is a problem... My bad that I did able to predict it.

Details:
Maven release plugin does simply bumb version to next. 10.4 -> 10.5, 10.1.4 -> 10.1.5.
So it means that in case of minor version bump , that is 90% of our versions, we will have validation failure immediately. But in reality we just waiting for new features to be merged.

Ideas to resolve:
1) release process need always to do extra step and make pom version to be 3 digits.
2) Convert to warning this specific case when version is 2 digits but we have no changes yes, all of them are miscellaneous, or bugs.
3) Convert to warning this specific case when version is 2 digits but we have no changes yes, all of them are miscellaneous. But as soon bug is merged do error.

Benefits of 1:
If CI green maintainers do release at any time
Brawback:
Extra automation required, may be extra activity in the middle to bump version up and down.

Benefits of 2 and 3:
- less commits of version bump
Drawback:
Maintainer should look at release notes before doing release. But this is always required, to make sure that titles are good and clear.

**Comment by nrmancuso** [comment]
> If CI green maintainers do release at any time, no brain

CI should be source of truth, we can always edit release notes but we cannot edit release/ tag number once the release is made. So I vote (1).

**Comment by stoyanK7** [comment]
I aggree with @nrmancuso and vote 1 too. Shall I proceed with a pull request, @romani ?

**Comment by romani** [comment]
Few more votes are required...
@rnveach , @pbludov , @strkkk .

**Comment by rnveach** [comment]
To me the problem is not very clear from @romani 's post as I have not been following all the details.

I am not against a 3 digit release. I guess I would say I would prefer to keep things as we have done just because that is what we have done.

I am not liking option 2 because it keeps things as a warnings and doesn't mention failing something somewhere. No one will constantly keep a check on it and it's going to be the case something is missed and a ops occurs.

I am ok with option 3 as long as it will be clear and a blocker that it won't accidentally be missed.

> Benefits of 2 and 3:
    less commits of version bump up and down.

If this is a blocker for a release, then it should be the first thing to fail in the line of release activities.

**Comment by romani** [comment]
before release, even commits are already merged, we can change title of issues to be used in release notes and we change labels.
after release, no changes can be done.

in our history I never did copy and usage of release notes blindly, I always tried to update issue titles to make them more readable by users who read release notes.
in our history there was bunch of cases where I changed label, as original labeling (usually done by last reviewer) was not correct.

so this this give me a hint  that it continue to exists. So we can not do releases without reviewing wording of them. So we need solution 2. As it is what we have now.

Or we change vision and state it like: "No need to polish issue titles, users do not read them that attentively" (kind of true) , "better to have peace of mind of maintainers and on green CI just trigger release and do not care about notes much"
in this case we need option 1. This option is very good if we start doing releases each week or on bug fix merge that is provided by outside contributor  - awesome contract with contributor, work on fix hard, after merge use it right away.

**Comment by romani** [comment]
We can use https://maven.apache.org/maven-release/maven-release-plugin/examples/non-interactive-release.html#using-system-properties

`-DdevelopmentVersion=......-SNAPSHOT` to define next version to be always patch number.

**Comment by romani** [comment]
After meditation on this, I think that (1) is very good way forward.

@stoyanK7 , we appreciate your help to make it.

**Review by romani — APPROVED** [review]
I am Merging it.
CI will not fail as drone is down :)

**Comment by stoyanK7** [comment]
I am on it.

**Pull request merged**
